### ISSUE TITLE
question：How to make an image compressed to a certain value

### ISSUE THREAD

**Issue by hunt-night** [AOMediaCodec/libavif#2492]
In my project, I want to compress images in bmp format, however, under the same encoder parameters the compression result of different photos varies a lot, is there any way to make the size of different images after encoding the same or the gap is small that is, the compression rate is fixed to a specific value

**Comment by vrabaud** [comment]
Some images are easier to compress than others. That's why, for a similar quality (which is the encoding parameter in libavif), you can have AVIF images of different sizes, though they have the same bmp size (bmp sizes are only dependent on the image dimensions so I guess your input images have the same dimensions).

What you will have to do is try different encoding qualities (through binary search to be faster) so that you get equivalent sizes in the end. The visual qualities will be different though.

**Comment by hunt-night** [comment]
Thanks for your reply
Can I assume that for different images of the same resolution in bmp format, it is possible that the original image itself will result in a compressed image with the same encoder parameters that will have one image with good compression and a high PSNR(computed by comparing the decompressed image to the original image), while the other image will have a relatively poor compression and a relatively low PSNR?

**Comment by vrabaud** [comment]
No, for given encoding parameters, the quality should be roughly the same for normal photos: the quality you set in https://github.com/AOMediaCodec/libavif/blob/dc2fe079b6e6d42e75f57fe053e8121972e90579/include/avif/avif.h#L1473 is responsible for the PSNR (there are a few other parameters here and there that can help too but this is the main one).

Then again, it depends on your data: even with a low quality parameter, a purely white image will be encoded with a good (maybe perfect) PSNR. What you describe is possible but for highly biased data, not random photos from the web.

**Comment by wantehchang** [comment]
The avifenc command has a `--target-size S` option that does the binary search Vincent mentioned automatically. (It was implemented by Yannis.)

The help message is "Set target file size in bytes (up to 7 times slower)". The binary search may perform up to 7 trial encodings (ceil(log2(100)) = 7).
